### PR TITLE
Add `hookContext.exceptionThrown`

### LINKED ISSUE #190: Incorrect currentBuild result on build failure inside @AfterStep hook
### Environment Information

- [2.249.3] Jenkins version
- [ 2.0.4] Plugin version
- [Windows 10 ] OS

### Description

I am trying to send some notification on build failure but currentBuild.result is returning null to me inside @AfterStep hook method.

Here is my hook method in step_watcher.groovy
@ AfterStep
void after(){
    println "Running after the ${hookContext.library} library's ${hookContext.step} step, currentBuild.currentResult= 
   ${currentBuild.currentResult}, currentBuild.result=${currentBuild.result}"
    //steps to send notification if build is failing
}
**Pipeline console output** -
[Pipeline] // stage
[JTE][@AfterStep - hooks/step_watcher.after]
[Pipeline] echo
Running after the git library's checkout step, currentBuild.currentResult= SUCCESS, currentBuild.result=null

Basically build is failing in checkout stage but it is still showing current build result as "SUCCESS".
image

### PR DESCRIPTION
**Pull request by steven-terrana**
# PR Details

fixes #190

the `currentBuild` variable available within Lifecycle Hook conditional execution closures doesn't indicate if the step or pipeline have failed. 

This is not a bug - but due to the fact that `currentBuild.result` isn't set to failure until the pipeline ends because it's possible the exception will be caught. 

To provide a mechanism for library authors to determine if steps have thrown an exception or if the pipeline is ending without catching an exception, a new field `exceptionThrown` is being added to `hookContext`. 

This would allow uses such as:

```groovy
@AfterStep({ hookContext.exceptionThrown })
void call(){
  println "only runs when the step threw an exception"
}
```

or

```groovy
@CleanUp({ hookContext.exceptionThrown })
void call(){
  println "only runs when pipeline is going to fail with an exception" 
}
```

## Types of Changes

- [ ] Added Unit Testing
- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] I have added the appropriate label for this PR
- [ ] If necessary, I have updated the documentation accordingly.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by cokieffebah** [comment]
Not for this PR but later ... should we have On(Step|Template)Exception Hooks?

**Comment by steven-terrana** [comment]
> Not for this PR but later ... should we have On(Step|Template)Exception Hooks?

we can let the people decide 🤷‍♂️.

File: docs/concepts/library-development/lifecycle-hooks.md [hunk 1]
```diff
@@ -26,28 +26,25 @@ Lifecycle Hook annotations can be placed on any method inside a step.
 
 Every step has an autowired `hookContext` variable which provides steps with relevant information about what triggered the hook.
 
-| Property              | Description                                                                                                                    |
-|-----------------------|--------------------------------------------------------------------------------------------------------------------------------|
-| `hookContext.library` | The name of the library that contributed the step associated with the step. Value is `null` for hooks not triggered by a step. |
-| `hookContext.step`    | The name of the Library Step that triggered the hook. Value is `null` for hooks not triggered by a step.                       |
+~~--8<-- "snippets/hookContext.md"~~
```
Comment (cokieffebah) [inline docs/concepts/library-development/lifecycle-hooks.md:29]: I think the ~~ are causing this to not be interpolated. the one in [docs/reference/autowired-variables.md](https://github.com/jenkinsci/templating-engine-plugin/pull/272/files#diff-5511781314f183198c3918e3b8c3e711003ee9d088c526ea81a622b9eaaac610) works

Comment (steven-terrana) [reply docs/concepts/library-development/lifecycle-hooks.md]: good catch - not sure how those `~`'s got there. thanks!

**Review by cokieffebah — APPROVED** [review]

**Pull request merged**